File: backend/app/api/v1/catalog.py
from urllib.parse import quote

from fastapi import APIRouter, Depends, File, Form, Response, UploadFile
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.response import ok
from app.core.security import CurrentUser, get_current_user
from app.schemas.catalog import (
    DataSourceCreateRequest,
    DataSourceUpdateRequest,
    ReportTemplateCreateRequest,
    ReportTemplateUpdateRequest,
    StatComponentCreateRequest,
    StatComponentUpdateRequest,
)
from app.services.catalog_service import CatalogService

router = APIRouter()


def get_service(
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
) -> CatalogService:
    return CatalogService(db, current_user)


@router.get("/templates")
def list_templates(service: CatalogService = Depends(get_service)):
    return ok(service.list_templates())


@router.post("/templates")
def create_template(req: ReportTemplateCreateRequest, service: CatalogService = Depends(get_service)):
    return ok(service.create_template(req))


@router.post("/templates/upload")
async def upload_template(
    file: UploadFile = File(...),
    name: str | None = Form(default=None),
    description: str = Form(default=""),
    status_value: str = Form(default="enabled", alias="status"),
    service: CatalogService = Depends(get_service),
):
    return ok(await service.upload_template(file, name, description=description, status_value=status_value))


@router.get("/templates/{template_id}/download")
def download_template(template_id: int, service: CatalogService = Depends(get_service)):
    filename, media_type, content = service.download_template(template_id)
    encoded_filename = quote(filename)
    return Response(
        content=content,
        media_type=media_type,
        headers={"Content-Disposition": f"attachment; filename*=UTF-8''{encoded_filename}"},
    )


@router.get("/templates/{template_id}/content")
def get_template_content(template_id: int, service: CatalogService = Depends(get_service)):
    return ok(service.get_template_content(template_id))


@router.put("/templates/{template_id}")
def update_template(template_id: int, req: ReportTemplateUpdateRequest, service: CatalogService = Depends(get_service)):
    return ok(service.update_template(template_id, req))


@router.delete("/templates/{template_id}")
def delete_template(template_id: int, service: CatalogService = Depends(get_service)):
    return ok(service.delete_template(template_id))


@router.get("/components")
def list_components(service: CatalogService = Depends(get_service)):
    return ok(service.list_components())


@router.post("/components")
def create_component(req: StatComponentCreateRequest, service: CatalogService = Depends(get_service)):
    return ok(service.create_component(req))


@router.put("/components/{component_id}")
def update_component(component_id: int, req: StatComponentUpdateRequest, service: CatalogService = Depends(get_service)):
    return ok(service.update_component(component_id, req))


@router.delete("/components/{component_id}")
def delete_component(component_id: int, service: CatalogService = Depends(get_service)):
    return ok(service.delete_component(component_id))


@router.get("/data-sources")
def list_data_sources(service: CatalogService = Depends(get_service)):
    return ok(service.list_data_sources())


@router.post("/data-sources")
def create_data_source(req: DataSourceCreateRequest, service: CatalogService = Depends(get_service)):
    return ok(service.create_data_source(req))


@router.put("/data-sources/{data_source_id}")
def update_data_source(data_source_id: int, req: DataSourceUpdateRequest, service: CatalogService = Depends(get_service)):
    return ok(service.update_data_source(data_source_id, req))


@router.delete("/data-sources/{data_source_id}")
def delete_data_source(data_source_id: int, service: CatalogService = Depends(get_service)):
    return ok(service.delete_data_source(data_source_id))

File: frontend/src/api/catalog.ts
import { apiDelete, apiGet, apiPost, apiPut } from './request'

export interface ReportTemplateItem {
  id: number
  name: string
  description: string
  content_json?: Record<string, unknown>
  original_filename?: string | null
  file_size?: number | null
  mime_type?: string | null
  status: string
  created_at: string
  updated_at: string
}

export interface StatComponentItem {
  id: number
  name: string
  component_type: string
  data_source: string
  usage: string
  config_json: Record<string, unknown>
  status: string
  created_at: string
  updated_at: string
}

export interface DataSourceItem {
  id: number
  name: string
  source_type: string
  address: string
  description: string
  config_json: Record<string, unknown>
  status: string
  created_at: string
  updated_at: string
}

export type TemplatePayload = Partial<Pick<ReportTemplateItem, 'name' | 'description' | 'content_json' | 'status'>>
export type ComponentPayload = Partial<Pick<StatComponentItem, 'name' | 'component_type' | 'data_source' | 'usage' | 'config_json' | 'status'>>
export type DataSourcePayload = Partial<Pick<DataSourceItem, 'name' | 'source_type' | 'address' | 'description' | 'config_json' | 'status'>>

export function listTemplates() {
  return apiGet<ReportTemplateItem[]>('/catalog/templates')
}

export function createTemplate(data: TemplatePayload) {
  return apiPost<ReportTemplateItem>('/catalog/templates', data)
}

export function uploadTemplate(file: File, data: { name?: string; description?: string; status?: string }) {
  const form = new FormData()
  form.append('file', file)
  if (data.name) form.append('name', data.name)
  form.append('description', data.description || '')
  form.append('status', data.status || 'enabled')
  return apiPost<ReportTemplateItem>('/catalog/templates/upload', form)
}

export interface ReportTemplateContent {
  id: number
  name: string
  original_filename: string
  html: string
}

export function getTemplateContent(id: number) {
  return apiGet<ReportTemplateContent>(`/catalog/templates/${id}/content`)
}

export async function downloadTemplate(id: number, filename: string) {
  const token = localStorage.getItem('report_access_token')
  const response = await fetch(`/api/v1/catalog/templates/${id}/download`, {
    headers: token ? { Authorization: `Bearer ${token}` } : {},
  })
  if (!response.ok) throw new Error(response.status === 404 ? '该模板没有可下载的 Word 文件' : '模板下载失败')
  const blob = await response.blob()
  const url = URL.createObjectURL(blob)
  const link = document.createElement('a')
  link.href = url
  link.download = filename || '报告模板.docx'
  document.body.appendChild(link)
  link.click()
  link.remove()
  URL.revokeObjectURL(url)
}

export function updateTemplate(id: number, data: TemplatePayload) {
  return apiPut<ReportTemplateItem>(`/catalog/templates/${id}`, data)
}

export function deleteTemplate(id: number) {
  return apiDelete<{ deleted: boolean }>(`/catalog/templates/${id}`)
}

export function listComponents() {
  return apiGet<StatComponentItem[]>('/catalog/components')
}

export function createComponent(data: ComponentPayload) {
  return apiPost<StatComponentItem>('/catalog/components', data)
}

export function updateComponent(id: number, data: ComponentPayload) {
  return apiPut<StatComponentItem>(`/catalog/components/${id}`, data)
}

export function deleteComponent(id: number) {
  return apiDelete<{ deleted: boolean }>(`/catalog/components/${id}`)
}

export function listDataSources() {
  return apiGet<DataSourceItem[]>('/catalog/data-sources')
}

export function createDataSource(data: DataSourcePayload) {
  return apiPost<DataSourceItem>('/catalog/data-sources', data)
}

export function updateDataSource(id: number, data: DataSourcePayload) {
  return apiPut<DataSourceItem>(`/catalog/data-sources/${id}`, data)
}

export function deleteDataSource(id: number) {
  return apiDelete<{ deleted: boolean }>(`/catalog/data-sources/${id}`)
}
